### PR TITLE
[Security] Tell about request_matcher

### LINKED ISSUE #11564: [Security] firewalls request_matcher no longer documented
`request_matcher` was present in https://symfony.com/doc/current/reference/configuration/security.html examples from 3.0 to 3.3 and then it disappeared.

Also it never appeared in https://symfony.com/doc/current/security/firewall_restriction.html

### PR DESCRIPTION
**Pull request by MatTheCat**
Fixes #11564

### CODE REVIEW

File: security/firewall_restriction.rst [hunk 1]
```diff
@@ -4,19 +4,77 @@
...
-restrict the initialization of a firewall against other options of the request.
+When using the Security component, firewalls will decide whether they handle a request based on the
+result of a request matcher: the first firewall matching the request will handle it.
+
+The last firewall can be configured without any matcher to handle every request it sees.
+
+Restricting by Service
+----------------------
+
+You can configure any service implementing ``Symfony\Component\HttpFoundation\RequestMatcherInterface``
```
Comment (OskarStark) [inline security/firewall_restriction.rst]:
```suggestion
You can configure any service implementing :class:`Symfony\\Component\\HttpFoundation\\RequestMatcherInterface`
```

File: security/firewall_restriction.rst [hunk 2]
```diff
@@ -4,19 +4,77 @@
...
+            'firewalls' => [
+                'secured_area' => [
+                    'request_matcher' => 'app.firewall.secured_area.request_matcher',
+                    // ...
+                ],
+            ],
+        ]);
+
+However in most cases you don’t need to create these matchers yourself as Symfony can do it for you based
+on the firewalls’ configuration.
```
Comment (OskarStark) [inline security/firewall_restriction.rst]:
```suggestion
on the firewall configuration.
```

File: security/firewall_restriction.rst [hunk 3]
```diff
@@ -4,19 +4,77 @@
 How to Restrict Firewalls to a Request
 ======================================
 
-When using the Security component, you can create firewalls that match certain request options.
-In most cases, matching against the URL is sufficient, but in special cases, you can further
-restrict the initialization of a firewall against other options of the request.
+When using the Security component, firewalls will decide whether they handle a request based on the
+result of a request matcher: the first firewall matching the request will handle it.
+
+The last firewall can be configured without any matcher to handle every request it sees.
```
Comment (OskarStark) [inline security/firewall_restriction.rst]:
```suggestion
The last firewall can be configured without any matcher to handle every incoming request.
```

**Comment by MatTheCat** [comment]
@OskarStark thanks for the suggestions! Do you know why the `Add suggestion to batch` is greyed? :thinking: 

suggestion

**Comment by OskarStark** [comment]
> @OskarStark thanks for the suggestions! Do you know why the Add suggestion to batch is greyed? 🤔

No, I don't know, but you can apply them one by one and squash the commits into one later. We can do this for you too when merging 👍

**Comment by MatTheCat** [comment]
Okay I'm not squashing yet as I guess there will be more reviews coming.

File: security/firewall_restriction.rst [hunk 4]
```diff
@@ -4,19 +4,77 @@
...
 
-When using the Security component, you can create firewalls that match certain request options.
-In most cases, matching against the URL is sufficient, but in special cases, you can further
-restrict the initialization of a firewall against other options of the request.
+When using the Security component, firewalls will decide whether they handle a request based on the
+result of a request matcher: the first firewall matching the request will handle it.
+
+The last firewall can be configured without any matcher to handle every incoming request.
+
+Restricting by Service
```
Comment (wouterj) [inline security/firewall_restriction.rst]: I would change the order and first documention "Restricting by Config" and then "Restricting by Service". I think the service one is not used that much, whereas the configuration is almost always used.

File: security/firewall_restriction.rst [hunk 5]
```diff
@@ -4,19 +4,77 @@
...
+        $container->loadFromExtension('security', [
+            'firewalls' => [
+                'secured_area' => [
+                    'request_matcher' => 'app.firewall.secured_area.request_matcher',
+                    // ...
+                ],
+            ],
+        ]);
+
+However in most cases you don’t need to create these matchers yourself as Symfony can do it for you based
```
Comment (wouterj) [inline security/firewall_restriction.rst]:
```suggestion
However in most cases you don't need to create these matchers yourself as Symfony can do it for you based
```

I don't think we use these nice typographically correct characters in the documentation (as most of us write the docs in their favorite IDE, so you'll always get the developer-single-quotes).

**Review by wouterj — APPROVED** [review]
Nice! Left two minor comments and other than that, this is ready to be merged!

**Review by OskarStark — APPROVED** [review]
After the comments

**Comment by MatTheCat** [comment]
Updated based on https://github.com/symfony/symfony-docs/pull/11574#discussion_r289199817 and https://github.com/symfony/symfony-docs/pull/11574#discussion_r289200081

I let you review once again!

**Comment by MatTheCat** [comment]
FYI I don't see the SymfonyCloud check mentionned in https://symfony.com/doc/current/contributing/documentation/overview.html#review-your-changes

**Comment by OskarStark** [comment]
No you can see the CI Step, sometimes it’s lacking a bit behind 😊

**Comment by MatTheCat** [comment]
Oh yeah I see it now :sweat_smile:

**Comment by MatTheCat** [comment]
Squashed and rebased.

**Comment by javiereguiluz** [comment]
Thanks for restoring this missing doc and for the overall improvement of this article.I Congrats on your first Symfony Docs contribution too!

**Pull request merged**